pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,441 @@
+            subs = dict(zip(pnames, psolns), t=intg.tcurr)
+
+            # Prepare any required gradients
+            if self._gradpinfo:
+                grads = np.rollaxis(grad_soln[idx], 2)[..., rgn]
+
+                # Interpolate the gradients to the points
+                grad_soln = soln_op @ grad_soln
+
+                # Transform from conservative to primitive gradients
+                pgrads = self.elementscls.grad_con_to_pri(soln, grads,
+                                                          self.cfg)
+
+                # Add them to the substitutions dictionary
+                for pname, idx in self._gradpinfo:
+                    for dim, grad in zip('xyz', pgrads[idx]):
+                        subs[f'grad_{pname}_{dim}'] = grad
+
+            for field, expr in self._expr.items():
+                if not isinstance(expr, dict):
+                    # Evaluate the expression at each point
+                    fun = npeval(expr, subs).T
+                    self.mesh_n[f'{d_str}/fields/{field}/values'] = fun
+                else:
+                    for sf, sexpr in expr.items():
+                        fun = npeval(sexpr, subs).T
+                        path = f'{d_str}/fields/{field}/values/{sf}'
+                        self.mesh_n[path] = fun
+
+    def _get_expr(self, expr, subs={}):
+        # Ensure the expression does not contain invalid characters
+        if not re.match(r'[A-Za-z0-9_ \t\n\r.,+\-*/%()]+$', expr):
+            raise ValueError('Invalid characters in expression')
+
+        # Substitute variables
+        if subs:
+            expr = re.sub(r'\b({0})\b'.format('|'.join(subs)),
+                          lambda m: str(subs[m[1]]), expr)
+
+        return expr
+
+    def _get_shape(self, name, nspts):
+        shapecls = subclass_where(BaseShape, name=name)
+        return shapecls(nspts, self.cfg)
+    
+    def _init_gradients(self, intg):
+        # Determine what gradients, if any, are required
+        g_pnames = set()
+        for field, expr in self._expr.items():
+            if expr and not isinstance(expr, dict):
+                g_pnames.update(re.findall(r'\bgrad_(.+?)_[xyz]\b', expr))
+            elif expr:
+                for sf, sexpr in expr.items():
+                    g_pnames.update(re.findall(r'\bgrad_(.+?)_[xyz]\b', sexpr))
+
+        privarmap = self.elementscls.privarmap[self.ndims]
+        self._gradpinfo = [(pname, privarmap.index(pname)) 
+                           for pname in g_pnames]
+
+    def __call__(self, intg):
+        if intg.nacptsteps % self.nsteps == 0:
+            comm, _, _ = get_comm_rank_root()
+
+            # Set file names
+            for sn, scfg in self._scene_cfg.items():
+                path = scfg['path_gen'].send(intg.tcurr)
+                self._add_scene[f'scenes/s_{sn}/image_name'] = path
+
+            # Set field expressions
+            self._eval_exprs(intg)
+
+            self.lib.ascent_publish(self.ascent_ptr, self.mesh_n)
```

Comment:
Check if we need to re-publish each time.

Reply:
Yep, seems like we do.